crypto/src/traits.rs

```diff
@@ -122,14 +126,44 @@ pub trait Authenticator:
 {
     type PubKey: VerifyingKey<Sig = Self>;
     type PrivKey: SigningKey<Sig = Self>;
+    type AggregateSig: AggregateAuthenticator<Sig = Self>;
 }
 
 /// Trait impl'd by a public / private key pair in asymmetric cryptography.
 ///
-pub trait KeyPair {
+pub trait KeyPair: Sized {
     type PubKey: VerifyingKey<PrivKey = Self::PrivKey>;
     type PrivKey: SigningKey<PubKey = Self::PubKey>;
     fn public(&'_ self) -> &'_ Self::PubKey;
-    fn private(self) -> Self::PrivKey;
+    fn private(&'_ self) -> &'_ Self::PrivKey;
     fn generate<R: CryptoRng + RngCore>(rng: &mut R) -> Self;
+    fn deterministic_generate(seed: &[u8], id: &[u8], domain: &[u8]) -> Result<Self, signature::Error>;
+}
+
+pub trait AggregateAuthenticator:
+    Display + Default + Serialize + DeserializeOwned + Send + Sync + 'static + Clone
+{
+    type Sig: Authenticator;
+    type PubKey: VerifyingKey<Sig = Self::Sig>;
+    type PrivKey: SigningKey<Sig = Self::Sig>;
+
+    /// Parse a key from its byte representation
+    fn aggregate(signatures: Vec<Self::Sig>) -> Result<Self, Error>;
+
+    fn add_signature(&mut self, signature: Self::Sig) -> Result<(), Error>;
+
+    fn add_aggregate(&mut self, signature: Self) -> Result<(), Error>;
+
+    /// Borrow a byte slice representing the serialized form of this key
+    fn verify(
+        &self,
+        pks: &[&<Self::Sig as Authenticator>::PubKey],
```

Comment:
So there's two approaches here: 
- conservative: require pubkeys to be stored in a contiguous array in memory, in which case you would want a `pks: &[<Self::Sig as Authenticator>::PubKey]`,
- liberal: require pubkeys to be something you can iterate over, without constraints on memory location.

In the liberal case, note that you could use an iterator type here. Consider using [`Borrow`](https://doc.rust-lang.org/std/borrow/trait.Borrow.html#examples) for its item type. 

There's one example of this [here](https://github.com/MystenLabs/mysten-infra/blob/7ce3f9bba93c93d87d02ec921ce7312c6f6dd0c4/crates/typed-store/src/rocks/mod.rs#L209-L212).

Reply:
I've changed everything to fit with the conservative approach!